### PR TITLE
Sidekiq Deployment Watchdog Worker

### PR DESCRIPTION
**Pull request by russellballestrini**
High Level Overview
===================

* Adds sidekiq to project
* Schedules a Deployment watchdog to run 30 seconds after creating the AutoDeployment

Heroku
========

Logged into UI and added redis as a "hobby dev" for now. Once provisioned, you can see the new redis url with the following command:

```
heroku config -a slashdeploy
```

Sidekiq expects you to set the following:

```
heroku config:set REDIS_PROVIDER=REDIS_URL -a slashdeploy
```
reference: https://github.com/mperham/sidekiq/wiki/Using-Redis#using-an-env-variable

Files Changed
===============

	modified:   Gemfile
	modified:   Gemfile.lock
	modified:   Procfile
	new file:   app/messages/auto_deployment_stuck_pending_message.rb
	modified:   app/models/auto_deployment.rb
	new file:   app/workers/deployment_watchdog_worker.rb
	new file:   bin/worker_deployment_watchdog
	modified:   lib/slashdeploy/service.rb
	modified:   spec/features/auto_deploy_spec.rb

### CODE REVIEW

File: app/workers/deployment_watchdog.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class DeploymentWatchdogWorker
+  include Sidekiq::Worker
+  sidekiq_options queue: 'deployment_watchdog'
+  def perform(auto_deployment_id)
+    auto_deployment = AutoDeployment.find_one(id: auto_deployment_id)
```
Comment (ejholmes) [inline app/workers/deployment_watchdog.rb]: Little more conventional to just do:

```ruby
AutoDeployment.find(auto_deployment_id)
```

File: app/workers/deployment_watchdog.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    # an inactive auto_deployment was either deployed, or superceded
+    # by another auto_deployment. Either way we exit without notifying.
+    return logger.debug "auto_deployment id #{auto_deployment.id} is inactive which means it was already deployed or superceded." unless auto_deployment.active?
+    if auto_deployment.pending?
+      logger.info "There was an issue with your deployment. Seems to be hung on #{auto_deployment.pending_statuses}"
+      message_class = AutoDeploymentStuckPendingMessage
+    elsif auto_deployment.failed?
```
Comment (ejholmes) [inline app/workers/deployment_watchdog.rb]: This case is already handled by SlashDeploy:

https://github.com/remind101/slashdeploy/blob/adea526cbfac35f7b001ae2984bc4f382299800e/spec/features/auto_deploy_spec.rb#L262-L265

File: app/workers/deployment_watchdog.rb (hunk 1 above)
Comment (russellballestrini) [reply app/workers/deployment_watchdog.rb]: @ejholmes can I still use `find_one` ?

Comment (ejholmes) [reply app/workers/deployment_watchdog.rb]: You can use it. In almost all Rails codebases (including this one) you'll see `.find(id)`, so I'd prefer consistency/convention if we can.

File: app/workers/deployment_watchdog_worker.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class DeploymentWatchdogWorker
+  include Sidekiq::Worker
+  sidekiq_options queue: 'deployment_watchdog'
+  def perform(auto_deployment_id)
+    slashdeploy = SlashDeploy::Service.new()
```
Comment (russellballestrini) [inline app/workers/deployment_watchdog_worker.rb]: Is this safe? Does this need args about the github/slack org?

File: spec/features/auto_deploy_spec.rb [hunk 4]
```diff
@@ -445,4 +448,54 @@
...
+    allow(github).to receive(:contents).and_return(config)
+
+    # prepare to receive a slack notification during push_event.
+    # these slack notification trigger when the config has required_context.
+    expect(slack).to receive(:direct_message).with \
+      slack_accounts(:david_baxterthehacker),
+      Slack::Message.new(text: ":wave: <@U012AB1AC>. I'll start a deployment of baxterthehacker/public-repo@0d1a26e to *production* for you once *ci/circleci* are passing.", attachments: [])
+
+    # our deployment watchdog worker should start with an empty queue.
+    assert_equal(0, DeploymentWatchdogWorker.jobs.size)
```
Comment (ejholmes) [inline spec/features/auto_deploy_spec.rb]: The convention in codebases that use rspec is to use `expect()`.

```ruby
expect(DeploymentWatchdogWorker.jobs.size).to eq 0
```

Let's keep it consistent with the rest of the codebase.

File: spec/features/auto_deploy_spec.rb [hunk 5]
```diff
@@ -1,5 +1,8 @@
 require 'rails_helper'
 
+require 'sidekiq/testing'
```
Comment (ejholmes) [inline spec/features/auto_deploy_spec.rb]: Can you put both of these lines in `spec/support/sidekiq.rb`? That will be required at boot time when running the tests.

File: lib/slashdeploy/service.rb [hunk 6]
```diff
@@ -45,6 +45,7 @@ def create_auto_deployment(environment, sha, user)
           auto_deployment.slack_account, \
           AutoDeploymentCreatedMessage, \
           auto_deployment: auto_deployment
+        DeploymentWatchdogWorker.perform_in(30.seconds, auto_deployment.id)
```
Comment (ejholmes) [inline lib/slashdeploy/service.rb]: Can you throw the `30.seconds` part into a constant somewhere?

Also, 30 seconds is way too short for what this PR is actually testing for; checking if an auto deployment is still pending. It's pretty normal for tests in CI to take 10-20 minutes, so we'd need something a little more reasonable (maybe like 1 hour?).

The 30 second value we came up with was for the scenario where a GitHub deployment is created, but nothing creates any deployment status events.

**Comment by ejholmes** [comment]
Also, if you merge. Can you rebase and squash to a single commit?

File: app/workers/deployment_watchdog_worker.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+class DeploymentWatchdogWorker
+  include Sidekiq::Worker
+  sidekiq_options queue: 'deployment_watchdog'
+
+  # default time to wait until our worker wakes up.
+  DEFAULT_DELAY = 1.hour
+
+  # create a class method to hardcode how we want to schedule this worker.
+  def self.schedule(auto_deployment_id)
```
Comment (ejholmes) [inline app/workers/deployment_watchdog_worker.rb]: I should mention, I love this change 👍

**Pull request merged**